File: composeApp/src/jvmMain/kotlin/org/sleepingcats/bridgecmdr/common/backup/model/Version4.kt
@file:OptIn(ExperimentalUuidApi::class)

package org.sleepingcats.bridgecmdr.common.backup.model

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import org.sleepingcats.bridgecmdr.common.backup.model.Version2.ColorScheme
import org.sleepingcats.bridgecmdr.common.setting.IconSize
import org.sleepingcats.bridgecmdr.common.setting.PowerOffTaps
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

object Version4 : BackupParser<Version5.Backup> {
  @Serializable
  data class Source(
    val id: Uuid,
    val title: String,
    val image: String? = null,
  )

  @Serializable
  data class Device(
    val id: Uuid,
    val driverId: Uuid,
    val title: String,
    val path: String,
  )

  @Serializable
  data class Tie(
    val id: Uuid,
    val sourceId: Uuid,
    val deviceId: Uuid,
    val inputChannel: Int,
    val outputVideoChannel: Int? = 0,
    val outputAudioChannel: Int? = 0,
  )

  @Serializable
  data class Layouts(
    val sources: List<Source>,
    val devices: List<Device>,
    val ties: List<Tie>,
  )

  @Serializable
  data class Settings(
    val iconSize: Int = 128,
    val colorScheme: ColorScheme = ColorScheme.NoPreference,
    val powerOnSwitchesAtStart: Boolean = false,
    val powerOffWhen: Version2.PowerOffTaps = Version2.PowerOffTaps.Single,
    val buttonOrder: List<Uuid> = emptyList(),
  )

  @Serializable
  data class Backup(
    val version: Int,
    val settings: Settings = Settings(),
    val layouts: Layouts,
  )

  fun upgradeBackup(backup: Backup) =
    Version5.Backup(
      version = 5,
      settings =
        Version5.Settings(
          appTheme = backup.settings.colorScheme.newValue,
          iconSize = IconSize.entries.find { entry -> entry.size == backup.settings.iconSize } ?: IconSize.Normal,
          powerOnDevicesAtStart = backup.settings.powerOnSwitchesAtStart,
          powerOffTaps = backup.settings.powerOffWhen.newValue,
          buttonOrder = backup.settings.buttonOrder,
        ),
      layouts = backup.layouts,
    )

  override fun parseBackup(raw: String): Version5.Backup =
    upgradeBackup(
      Json.decodeFromString<Backup>(raw).also {
        check(it.version == 4) { "Wrong backup version, expected 4" }
      },
    )
}

File: src/renderer/services/backup/formats/version4.ts
import { z } from 'zod'
import { ButtonOrder, ColorScheme, IconSize, PowerOffTaps } from '../../settings'

/** v4 general document header. */
export const DocHeader = z.object({
  id: z.string().uuid().toUpperCase()
})

/**
 * Layout v4, will be used in Export v4 but
 * may be reused by later versions.
 */
export type Layouts = z.output<typeof Layouts>
export const Layouts = z.object({
  sources: z.array(
    DocHeader.extend({
      title: z.string().min(1),
      image: z.string().nullish()
    })
  ),
  devices: z.array(
    DocHeader.extend({
      driverId: z.string().uuid().toUpperCase(),
      title: z.string(),
      path: z.string()
    })
  ),
  ties: z.array(
    DocHeader.extend({
      sourceId: z.string().uuid().toUpperCase(),
      deviceId: z.string().uuid().toUpperCase(),
      inputChannel: z.number().int(),
      outputVideoChannel: z.number().int().nullish(),
      outputAudioChannel: z.number().int().nullish()
    })
  )
})

/**
 * Settings v4, will be used in Export v4 but
 * may be reused by later versions.
 */
export type Settings = z.output<typeof Settings>
export const Settings = z
  .object({
    iconSize: IconSize.optional(),
    colorScheme: ColorScheme.optional(),
    powerOnSwitchesAtStart: z.boolean().optional(),
    powerOffWhen: PowerOffTaps.optional(),
    buttonOrder: ButtonOrder.optional()
  })
  .nullish()

/**
 * Backup parser for v4.
 */
export const Backup = z
  .object({
    version: z.literal(4),
    settings: Settings,
    layouts: Layouts
  })
  .transform((payload) => ({
    version: 'normalized' as const,
    settings: payload.settings,
    layouts: {
      sources: payload.layouts.sources.map(({ id, image = null, ...source }) => ({
        _id: id,
        order: 0,
        image,
        ...source
      })),
      devices: payload.layouts.devices.map(({ id, ...device }) => ({
        _id: id,
        ...device
      })),
      ties: payload.layouts.ties.map(({ id, outputVideoChannel, outputAudioChannel, ...tie }) => ({
        _id: id,
        ...tie,
        outputChannels: {
          video: outputVideoChannel ?? undefined,
          audio: outputAudioChannel ?? undefined
        }
      }))
    }
  }))

/** Backup parser output. */
export type Import = z.output<typeof Backup>

export type Export = z.input<typeof Backup>
